nibabel/filebasedimages.py

```diff
@@ -0,0 +1,537 @@
+       * instance_to_filename(img, fname) - save ``img`` instance to
+         filename ``fname``.
+
+    It also has a ``header`` - some standard set of meta-data that is specific to
+    the image format, and ``extra`` - a dictionary container for any other
+    metadata.
+
+    You cannot slice an image, and trying to slice an image generates an
+    informative TypeError.
+
+
+    There are several ways of writing data.
+    =======================================
+
+    There is the usual way, which is the default::
+
+        img.to_filename(fname)
+
+    and that is, to take the data encapsulated by the image and cast it to
+    the datatype the header expects, setting any available header scaling
+    into the header to help the data match.
+
+    You can load the data into an image from file with::
+
+       img.from_filename(fname)
+
+    The image stores its associated files in its ``file_map`` attribute.  In order
+    to just save an image, for which you know there is an associated filename, or
+    other storage, you can do::
+
+       img.to_file_map()
+
+    You can get the data out again with::
+
+        img.get_data()
+
+    Less commonly, for some image types that support it, you might want to
+    fetch out the unscaled array via the object containing the data::
+
+        unscaled_data = img.dataoobj.get_unscaled()
+
+    Analyze-type images (including nifti) support this, but others may not
+    (MINC, for example).
+
+    Sometimes you might to avoid any loss of precision by making the
+    data type the same as the input::
+
+        hdr = img.header
+        hdr.set_data_dtype(data.dtype)
+        img.to_filename(fname)
+
+    Files interface
+    ===============
+
+    The image has an attribute ``file_map``.  This is a mapping, that has keys
+    corresponding to the file types that an image needs for storage.  For
+    example, the Analyze data format needs an ``image`` and a ``header``
+    file type for storage:
+
+       >>> import nibabel as nib
+       >>> data = np.arange(24, dtype='f4').reshape((2,3,4))
+       >>> img = nib.AnalyzeImage(data, np.eye(4))
+       >>> sorted(img.file_map)
+       ['header', 'image']
+
+    The values of ``file_map`` are not in fact files but objects with
+    attributes ``filename``, ``fileobj`` and ``pos``.
+
+    The reason for this interface, is that the contents of files has to
+    contain enough information so that an existing image instance can save
+    itself back to the files pointed to in ``file_map``.  When a file holder
+    holds active file-like objects, then these may be affected by the
+    initial file read; in this case, the contains file-like objects need to
+    carry the position at which a write (with ``to_files``) should place the
+    data.  The ``file_map`` contents should therefore be such, that this will
+    work:
+    '''
+    files_types = (('image', None),)
+    alternate_exts = ()  # Modified by @ImageOpener.register_ext_from_image
```

Comment:
`sizeof_hdr` is on the header, [`_meta_sniff_len`](https://github.com/nipy/nibabel/pull/364/files#diff-1053478051df4b99b0c69a5365f5f0beL326) is on the image and used in [`path_maybe_image`](https://github.com/nipy/nibabel/pull/364/files#diff-496e321b697d515cfea0150d3c97dfb5R530) without being checked to exist, first. (It does check whether the header class defines `may_contain_header`, but that's not tied to presence of `_meta_sniff_len`, AFAICT.)

`files_types` must also be defined by subclasses, correct? I think its purpose here is documentary. But the point of my comment was that `alternate_exts` has been removed from `SpatialImage`.